File: Ukemochi-Scripting/UkemochiEngine/CoreModule/GameObject.cs
/* Start Header
 *****************************************************************/
/*!
\file	Transform.cs
\par	Ukemochi
\author WONG JUN YU, Kean, junyukean.wong, 2301234
\par	junyukean.wong\@digipen.edu
\par	Course: CSD2400/CSD2401
\date	31/10/24
\brief	This files is responsible for handling all the gameobject's components, and the first component is always the transform

Copyright (C) 2024 DigiPen Institute of Technology.
Reproduction or disclosure of this file or its contents without the
prior written consent of DigiPen Institute of Technology is prohibited.
*/
/* End Header
 *******************************************************************/

using System;

// For the love of all things, here's me writing down my thoughts
// C++ will InstantiateClass of this class as an Object when a user of the editor calls add gameobject
// This object holds a list of components, and the first component is always the transform
// Follow the same principle as above, when a user call add component from the editor, the engine will instantiate the class of that component

namespace Ukemochi
{
    public class GameObject : Object
    {
        // ==================== PROPERTIES ====================
        public Transform transform { get; private set; } // Read-only
        
        // ==================== METHODS =======================
        public bool HasComponent<T>() where T : Component
        {
            Type componentType = typeof(T);
            return EngineInterop.HasComponent(GetInstanceID(), componentType);
        }

        public T GetComponent<T>() where T : Component, new()
        {
            if (!HasComponent<T>())
                return null;
            if (transform == null)
            {
                transform = GetComponent<Transform>();
                transform.SetGameObject(this);
                _id = GetInstanceID();
            }

            T component = new T();
            component._id = GetInstanceID();
            component.SetGameObject(this);
            return component;
        }
        
        public static GameObject FindWithTag(string tag)
        {
            EngineInterop.GetObjectByTag(tag, out ulong id);
            if (id == 999999)
                return null;
            
            var obj = new GameObject();
            obj._id = id;
            obj.transform = new Transform();
            obj.transform.SetGameObject(obj);
            obj.transform._id = id;
            return obj;
        }

        public override string ToString()
        {
            return base.ToString();
        }
    }
}

File: Ukemochi-Engine/src/Ukemochi-Engine/Factory/GameObject.cpp
/* Start Header ************************************************************************/
/*!
\file       GameObject.cpp
\author     Pek Jun Kai Gerald, p.junkaigerald, 2301334, p.junkaigerald\@digipen.edu (50%)
\co-authors Wong Jun Yu, Kean, junyukean.wong, 2301234, junyukean.wong\@digipen.edu (50%)
\date       Sept 29, 2024
\brief      This file contains the declaration of the GameObject class. This class is responsible for
        creating and managing game objects. This applies to the C# side as well. It reflects the gameobject instance
        on the C# side of things.

Copyright (C) 2024 DigiPen Institute of Technology.
Reproduction or disclosure of this file or its contents without the
prior written consent of DigiPen Institute of Technology is prohibited.
*/
/* End Header **************************************************************************/

#include "PreCompile.h"
#include "GameObject.h"
using namespace Ukemochi;

/**
 * @brief Construct a new GameObject object
 * 
 * @param id The entity ID
 * @param name The name of the game object
 * @param tag The tag of the game object
 */
GameObject::GameObject(EntityID id, std::string name, std::string tag): m_Name(std::move(name)),
                                                                        m_Tag(std::move(tag)), m_InstanceID(id)
{
    // m_ManagedInstance = ScriptingEngine::GetInstance().InstantiateClass("GameObject");
    //
    // m_ManagedInstanceHandle = mono_gchandle_new_v2(m_ManagedInstance, true);
    // ScriptingEngine::SetMonoFieldValueULL(m_ManagedInstance, "_id", &m_InstanceID);
    // ScriptingEngine::SetMonoFieldValueString(m_ManagedInstance, "name", m_Name);
}

/**
 * @brief Copy constructor
 * 
 * @param other The game object to copy
 */
GameObject& GameObject::operator=(const GameObject& other)
{
    m_Name = other.m_Name;
    m_Tag = other.m_Tag;
    m_InstanceID = other.m_InstanceID;
    
    return *this;
}

/**
 * @brief Get the Entity ID of the game object
 * @return EntityID The instance ID of the game object 
 */
EntityID GameObject::GetInstanceID() const
{
    return m_InstanceID;
}

/**
 * @brief Get the active state of the gameobject
 * @return bool the active state of the game object
 */
bool GameObject::GetActive() const
{
    return m_isActive;
}
